prediction_market_agent/agents/microchain_agent/app_nft_game.py

```diff
@@ -0,0 +1,169 @@
+        SendPaidMessageToAnotherAgent()(nft_agent.wallet_address, message)
+        st.success("Message sent and will be processed soon!")
+
+
+def parse_function_and_body(
+    role: t.Literal["user", "assistant", "system"], message: str
+) -> t.Tuple[str | None, str | None]:
+    message = message.strip()
+
+    if role == "assistant":
+        parsed_function = message.split("(")[0]
+        parsed_body = message.split("(")[1].rsplit(")")[0]
+    elif role == "user":
+        parsed_function = "Response"
+        parsed_body = message
+    else:
+        parsed_function = None
+        parsed_body = None
+
+    return parsed_function, parsed_body
+
+
+def customized_chat_message(
+    role: t.Literal["user", "assistant", "system"],
+    message: str,
+    created_at: DatetimeUTC,
+) -> None:
+    parsed_function, parsed_body = parse_function_and_body(role, message)
+    if parsed_function is None:
+        return
+
+    match parsed_function:
+        case "Response":
+            icon = "✔️"
+        case ReceiveMessage.__name__:
+            icon = "👤"
+        case BroadcastPublicMessageToHumans.__name__:
+            icon = "📣"
+        case SendPaidMessageToAnotherAgent.__name__:
+            icon = "💸"
+        case _:
+            icon = "🤖"
+
+    with st.chat_message(icon):
+        if parsed_function:
+            st.markdown(f"**{parsed_function}**")
+        st.write(created_at.strftime("%Y-%m-%d %H:%M:%S"))
+        if message:
+            st.markdown(parsed_body)
+
+
+def show_function_calls_part(nft_agent: NFTAgent) -> None:
+    st.markdown(f"""### Agent's actions""")
+
+    with st.spinner("Loading agent's actions..."):
+        calls = long_term_memory_table_handler(nft_agent.identifier).search()
+
+    if not calls:
+        st.markdown("No actions yet.")
+        return
+
+    for item in calls:
+        if item.metadata_dict is None:
+            continue
+        customized_chat_message(
+            item.metadata_dict["role"],
+            item.metadata_dict["content"],
+            item.datetime_,
+        )
+
+
+def show_about_agent_part(nft_agent: NFTAgent) -> None:
+    system_prompt = prompt_table_handler(nft_agent.identifier).fetch_latest_prompt()
+    xdai_balance = get_balances(nft_agent.wallet_address).xdai
+    st.markdown(
+        f"""### {nft_agent.name}
+
+Currently holds <span style='font-size: 1.1em;'><strong>{xdai_balance:.2f} xDAI</strong></span>.
+
+---
+""",
+        unsafe_allow_html=True,
+    )
+    st.text_area(
+        f"{nft_agent.name}'s system prompt",
+        value=system_prompt.prompt if system_prompt else "No system prompt yet.",
+        disabled=True,
+    )
+
+
+def get_agent_page(nft_agent: NFTAgent) -> t.Callable[[], None]:
+    def page() -> None:
+        left, _, right = st.columns([0.3, 0.05, 0.65])
+
+        with left:
+            show_about_agent_part(nft_agent)
+
+        with right:
+            send_message_part(nft_agent)
+            show_function_calls_part(nft_agent)
```

Comment:
Wondering how often this will be triggered - I guess that we would need to refresh the page?
Maybe using `st.cache_data` could be helpful?

Reply:
Hmm good point, I'll check if we can auto-refresh this part somehow.

Reply:
Done! Works like a charm.